plutus-ledger/src/Ledger/Validation.hs

```diff
@@ -0,0 +1,264 @@
+
+* We cannot represent different eras - everything is 'AlonzoEra'.
+* There is no handling of epoch boundaries, rewards, etc.
+* The block size is unlimited - we simply take all transactions from the
+  mempool when we make a block. There is however a limit on the size of
+  individual transactions.
+* We use the standard ledger cryptography everywhere ('StandardCrypto').
+  This could be replaced by "NoCrypto" for faster validation.
+
+-}
+
+{-| State of the ledger with configuration, mempool, and the blockchain.
+-}
+data EmulatedLedgerState =
+  EmulatedLedgerState
+    { _ledgerEnv      :: MempoolEnv EmulatorEra
+    , _memPoolState   :: MempoolState EmulatorEra
+    , _currentBlock   :: EmulatorBlock
+    , _previousBlocks :: [EmulatorBlock]
+    }
+
+makeLenses ''EmulatedLedgerState
+
+{-| Increase the slot number by one
+-}
+nextSlot :: EmulatedLedgerState -> EmulatedLedgerState
+nextSlot = over ledgerEnv f where
+  f l@LedgerEnv{ledgerSlotNo=oldSlot} = l{ledgerSlotNo = succ oldSlot}
+
+{-| Set the slot number
+-}
+setSlot :: SlotNo -> EmulatedLedgerState -> EmulatedLedgerState
+setSlot sl = over ledgerEnv (\l -> l{ledgerSlotNo=sl})
+
+{-| Make a block with all transactions that have been validated in the
+current block, add the block to the blockchain, and empty the current block.
+-}
+makeBlock :: EmulatedLedgerState -> EmulatedLedgerState
+makeBlock state =
+  state
+    & currentBlock .~ []
+    & over previousBlocks ((:) (reverse $ state ^. currentBlock))
+
+{-| Initial ledger state for a distribution
+-}
+initialState :: EmulatedLedgerState
+initialState = EmulatedLedgerState
+  { _ledgerEnv = C.Ledger.mkMempoolEnv emulatorNes 0
+  , _memPoolState = C.Ledger.mkMempoolState emulatorNes
+  , _currentBlock = []
+  , _previousBlocks = []
+  }
+
+utxoEnv :: SlotNo -> UtxoEnv EmulatorEra
+utxoEnv slotNo = UtxoEnv slotNo emulatorPParams mempty (C.Ledger.GenDelegs mempty)
+
+applyTx ::
+  EmulatedLedgerState ->
+  Tx EmulatorEra ->
+  Either ValidationError (EmulatedLedgerState, Validated (Tx EmulatorEra))
+applyTx oldState@EmulatedLedgerState{_ledgerEnv, _memPoolState} tx = do
+  (newMempool, vtx) <- first (CardanoLedgerValidationError . show) (C.Ledger.applyTx emulatorGlobals _ledgerEnv _memPoolState tx)
+  return (oldState & memPoolState .~ newMempool & over currentBlock ((:) vtx), vtx)
+
+
+emulatorNetworkId :: C.Api.NetworkId
+emulatorNetworkId = C.Api.Testnet $ C.Api.NetworkMagic 1
+
+-- TODO: the larger maxTxSize should only be used when needed.
+genesisDefaultsWithBigMaxTxSize :: C.Ledger.ShelleyGenesis EmulatorEra
+genesisDefaultsWithBigMaxTxSize = shelleyGenesisDefaults {
+  C.Ledger.sgProtocolParams = (C.Ledger.sgProtocolParams shelleyGenesisDefaults) {
+    C.Ledger._maxTxSize = 256 * 1024
+  }
+}
+
+{-| A sensible default 'Globals' value for the emulator
+-}
+emulatorGlobals :: Globals
+emulatorGlobals = mkShelleyGlobals
+  genesisDefaultsWithBigMaxTxSize
+  (fixedEpochInfo (EpochSize 432000) (mkSlotLength 8)) -- ?
```

Comment:
What's the "?" for?

Reply:
This was already in Jann's code. I don't think it matters for our use cases.